pubsub/rabbitmq/rabbitmq.go

```diff
@@ -0,0 +1,176 @@
+	meta, err := createMetadata(metadata)
+	if err != nil {
+		return err
+	}
+
+	r.metadata = meta
+
+	conn, err := amqp.Dial(meta.host)
+	if err != nil {
+		return err
+	}
+
+	ch, err := conn.Channel()
+	if err != nil {
+		return err
+	}
+
+	r.connection = conn
+	r.channel = ch
+	return nil
+}
+
+func (r *rabbitMQ) Publish(req *pubsub.PublishRequest) error {
+	err := r.ensureExchangeDeclared(req.Topic)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("%s publishing message to topic '%s'", logMessagePrefix, req.Topic)
+
+	err = r.channel.Publish(req.Topic, "", false, false, amqp.Publishing{
+		ContentType:  "text/plain",
+		Body:         req.Data,
+		DeliveryMode: r.metadata.deliveryMode,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *rabbitMQ) Subscribe(req pubsub.SubscribeRequest, handler func(msg *pubsub.NewMessage) error) error {
+
+	err := r.ensureExchangeDeclared(req.Topic)
+	if err != nil {
+		return err
+	}
+
+	queueName := fmt.Sprintf("%s-%s", r.metadata.consumerID, req.Topic)
+
+	log.Debugf("%s declaring queue '%s'", logMessagePrefix, queueName)
+	q, err := r.channel.QueueDeclare(queueName, r.metadata.durable, r.metadata.deleteWhenUnused, true, false, nil)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("%s binding queue '%s' to exchange '%s'", logMessagePrefix, q.Name, req.Topic)
+	err = r.channel.QueueBind(q.Name, "", req.Topic, false, nil)
+	if err != nil {
+		return err
+	}
+
+	msgs, err := r.channel.Consume(
+		q.Name,
+		queueName,           // consumerId
+		r.metadata.autoAck,  // autoAck
+		!r.metadata.durable, // exclusive
+		false,               // noLocal
+		false,               // noWait
+		nil,
+	)
+
+	if err != nil {
+		return err
+	}
+
+	go r.listenMessages(msgs, req.Topic, handler)
+
+	return nil
+}
+
+func (r *rabbitMQ) listenMessages(msgs <-chan amqp.Delivery, topic string, handler func(msg *pubsub.NewMessage) error) {
+	for d := range msgs {
+		r.handleMessage(d, topic, handler)
+	}
+}
+
+func (r *rabbitMQ) handleMessage(d amqp.Delivery, topic string, handler func(msg *pubsub.NewMessage) error) {
+	pubsubMsg := &pubsub.NewMessage{
+		Data:  d.Body,
+		Topic: topic,
+	}
+
+	err := handler(pubsubMsg)
+	if err != nil {
+		log.Errorf("%s error handling message from topic '%s', %s", errorMessagePrefix, topic, err)
+	}
+
+	// if message is not auto acked we need to ack/nack
+	if !r.metadata.autoAck {
+
+		if err != nil {
+			requeue := r.metadata.requeueInFailure && !d.Redelivered
+
+			log.Debugf("%s nacking message '%s' from topic '%s', requeue=%t", logMessagePrefix, d.MessageId, topic, requeue)
+			if err = r.channel.Nack(d.DeliveryTag, false, requeue); err != nil {
+				log.Errorf("%s error nacking message '%s' from topic '%s', %s", d.MessageId, logMessagePrefix, topic, err)
+			}
+		} else {
+			log.Debugf("%s acking message '%s' from topic '%s'", logMessagePrefix, d.MessageId, topic)
+			if err = r.channel.Ack(d.DeliveryTag, false); err != nil {

```

Comment:
Should only ack here if the `handler` method from line 139 returned nil

Reply:
Hi @yaron2,

Thanks for the review.
I think I am missing something here. Thinking about the scenarios:

1. Auto ack == true
  In this case we don't ack or nack since the message was already acked during dequeue. This option was by default true, but I am changed to false to better handle scenarios where daprd is running before the consumer is ready.

2. AutoAck == false and err != nil
  Flow goes to lines:
    - 141 (log error)
    - 147 (block, including logging and nack)

3. AutoAck == false and err == nil
  Flow goes to line:
    - 155 (block, including logging and ack)